[primTorch] Minor improvements to doc and impl of `gaussian_nll_loss`

nkaretnikov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #85612

Fixes #53392.

cc @ezyang @mruberry @ngimel @Lezcano @peterbell10

nkaretnikov committed:
[primTorch] Add ref for `gaussian_nll_loss`, add error inputs

[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Add ref for `gaussian_nll_loss`, add error inputs"




[ghstack-poisoned]

nkaretnikov commented:
Note: see https://github.com/pytorch/pytorch/issues/53964/ and https://github.com/pytorch/pytorch/pull/56469 for context on the size checks and clamping before the loss calculation.

nkaretnikov committed:
Update on "[primTorch] Add ref for `gaussian_nll_loss`, add error inputs"




[ghstack-poisoned]

nkaretnikov committed:
Update on "[primTorch] Add ref for `gaussian_nll_loss`, add error inputs"




[ghstack-poisoned]

lezcano commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -147,6 +148,82 @@ def elu(
+    if full:
+        # torch.log doesn't support scalars, so multiply by a tensor.
+        # Also, avoid inplace add.
+        loss = loss + (0.5 * torch.log(2 * torch.pi * torch.ones_like(loss)))
```
Why did you deviate from the implementation in core here? You went from calling no kernels and just a few python ops, to calling 5 kernels...
```suggestion
        loss = loss + 0.5 * math.log(2 * math.pi)
```
Be careful with the cahnges you introduce.

lezcano commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -609,23 +686,6 @@ def relu6(a: TensorLikeType, inplace: bool = False) -> TensorLikeType:
     return refs.nn.functional.hardtanh(a, 0, 6)
 
 
-@register_decomposition(torch.ops.aten.glu)
-@elementwise_type_promotion_wrapper(
-    type_promoting_args=("a",),
-    type_promotion_kind=utils.ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
-)
-@out_wrapper()
-def glu(a: TensorLikeType, dim: int = -1) -> TensorLikeType:
-    dim = utils.canonicalize_dims(a.ndim, dim)
-    check(
-        a.shape[dim] % 2 == 0,
-        lambda: f"Halving dimension must be even, but dimension {dim} is size {a.shape[dim]}",
-    )
-    b, c = torch.tensor_split(a, 2, dim)
-
-    return b * torch.sigmoid(c)
-
-
```
What does the glu have to do with this PR?

lezcano reviewed (commented):
This is an interesting one. This function is already implemented in Python in core, so it's not clear to me whether we want to re-implement it in PrimTorch. The only benefit I see in doing this is that we can implement some promotion rules for it, and the fact that this way we would have all our implementations in the same place... WDYT @mruberry? Do

lezcano commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -147,6 +148,82 @@ def elu(
+    # Check validity of reduction mode
+    _check_reduction_value(reduction)
+
+    # Entries of var must be non-negative
+    if torch.any(var < 0).item():
+        raise ValueError("var has negative entry/entries")
```
We should remove this from here and from core and change it by saying in the docs that if `var < 0`, then the implementation is undefined.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -7163,6 +7163,32 @@ def gen_shape_kwargs():
     for input, target, var, kwargs in gen_shape_kwargs():
         yield SampleInput(input, args=(target, var, ), kwargs=kwargs)
 
+def error_inputs_gaussian_nll_loss(op_info, device, **kwargs):
```
Nice error input addition

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -7163,6 +7163,32 @@ def gen_shape_kwargs():
+    make_input = partial(make_tensor, device=device, dtype=torch.float32)
+
+    samples = (
+        # input, args, kwargs, error_type, error_regex
```
Good comments explaining the structure of the sequence and the errors being checked

mruberry commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -147,6 +148,82 @@ def elu(
+    diff = input - target
+    loss = 0.5 * (torch.log(var) + diff * diff / var)
+    if full:
+        # torch.log doesn't support scalars, so multiply by a tensor.
```
Ooops, @lezcano suggests the same thing below

nkaretnikov committed:
Update on "[primTorch] Add ref for `gaussian_nll_loss`, add error inputs"




[ghstack-poisoned]

nkaretnikov commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -147,6 +148,82 @@ def elu(
+    reduction: str = "mean",
+) -> TensorLikeType:
+    # Check var size
+    # If var.size == input.size, the case is heteroscedastic and no further
```
rewrote the comment, let me know whether the new one is fine

mruberry commented on test/test_nn.py:
```diff
@@ -6816,7 +6816,7 @@ def test_gaussian_nll_loss_reduction_modes(self):
         self.assertEqual(torch.mean(component_wise_loss),
                          F.gaussian_nll_loss(input, target, var, reduction='mean'))
-        with self.assertRaisesRegex(ValueError, 'is not valid'):
+        with self.assertRaisesRegex(ValueError, 'total is not a valid value for reduction'):
```
I think this should be an ErrorInput?

mruberry replied:
(actually, it seems so redundant with the `abc` ErrorInput below maybe it can just be deleted?)

mruberry commented on test/test_nn.py:
```diff
@@ -6844,9 +6844,6 @@ def test_gaussian_nll_loss_args(self):
             target = torch.randn(3, 5)
             var = torch.ones(3, 3)
             torch.nn.functional.gaussian_nll_loss(input, target, var)
-        with self.assertRaisesRegex(ValueError, 'var has negative entry/entries'):
```
I think the previous assertion can also be an ErrorInput and this test removed?

mruberry replied:
While I suspect this test can be removed entirely, `test_gaussian_nll_loss_broadcasting` and `test_gaussian_nll_loss_reduction_modes` may need to stick around for now -- but we should add a comment that they should eventually be made redundant with the OpInfo

mruberry commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -147,6 +149,64 @@ def elu(
+    # To support broadcasting, the following sub-cases are allowed in the
+    # homoscedastic case.  With all other sizes being the same as input (when
+    # comparing from the outermost dimension), var must either have:
+    # 1) one fewer dimension or
```
More precisely, `input` and `var` represent the expectations and variances of a multivariate normal (AKA "Gaussian") distribution. The operation does NOT support traditional broadcasting between input and var. It does support

1. Input and var being the same shape
2. Input and var being the same shape except var has an innermost dimension of length one, which is then broadcast
3. Input and var being the same shape except var is missing the corresponding innermost dimension, in which case var is unsqueezed to create that innermost dimension and then it's broadcast as in (2)

If the variance is constant across distributions then a short-hand this allows to represent that is either:

1. `input` and `var` have the same shape, except `var` is missing the innermost dimension

nkaretnikov replied:
> Do input and var represent the expectations and variances of a single multivariate normal (AKA "Gaussian") distribution? Or are these the expectations and variances of multiple independent distributions, like one per each class?

I've looked at two arXiv papers linked from the original issue https://github.com/pytorch/pytorch/issues/48520 and the related PRs/issues, but I'm still not sure and would rather not change this part of the doc. I did remove the description from the functional op to have one source of truth.

mruberry replied:
I understand but I think we should still do our best to describe the mathematics of what the operation does in this case -- that should lead to some interpretation?

mruberry commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -147,6 +149,64 @@ def elu(
+    diff = input - target
+    loss = 0.5 * (torch.log(var) + diff * diff / var)
+    if full:
+        # Avoid inplace add
```
"Avoid" -> "Avoids"

nkaretnikov commented:
reminder to close this one once we fix the docs: https://github.com/pytorch/pytorch/issues/53392

nkaretnikov committed:
Update on "[primTorch] Minor improvements to doc and impl of `gaussian_nll_loss`"




[ghstack-poisoned]

nkaretnikov commented on torch/nn/functional.py:
```diff
@@ -2777,8 +2777,10 @@ def gaussian_nll_loss(
         target: sample from the Gaussian distribution.
-        var: tensor of positive variance(s), one for each of the expectations
-            in the input (heteroscedastic), or a single one (homoscedastic).
+        var: same shape as the input, or same shape as the input but with the
```
I just removed the doc from the functional part. See the above comment WRT the Gaussian bit.

nkaretnikov committed:
Update on "[primTorch] Minor improvements to doc and impl of `gaussian_nll_loss`"


Fixes #53392.

[ghstack-poisoned]

nkaretnikov commented on torch/nn/functional.py:
```diff
@@ -2817,14 +2817,15 @@ def gaussian_nll_loss(
         raise ValueError(reduction + " is not a valid value for reduction")
 
     # Clamp for stability
-    var = var.clone()
```
not sure what was the purpose of cloning here since the variable is shadowed later anyway

nkaretnikov replied:
Now I know why. It's either the original behavior (with `no_grad` and inplace clamp) or just this (without any context): `var = var.clamp(min=eps)`. I couldn't find any other code that wouldn't break the following tests:

```
python -m pytest test/test_modules.py -k GaussianNLLLoss -vvv
python -m pytest test/test_ops_gradients.py -k gaussian_nll_loss -vvv
```

[This](https://github.com/pytorch/pytorch/issues/53964#issuecomment-798067924) claims that doing it without `no_grad` will "cause divergence," but the tests pass locally.

lezcano replied:
any thoughts @albanD ? Clamping without `no_grad` LGTM, but I know we've historically done it with the no_grad...

albanD replied:
clamping with `no_grad` will zero out a bunch of gradients. We definitely don't want that!

lezcano replied:
What about figuring out what's the value we want this function to take at `var = 0` and return that? Although we would need to somehow deal with the gradients as well...

nkaretnikov replied:
Note: @lezcano told me offline there are plans (IIUC) to have a "framework" that would help with numerical issues like this one, so postponing for now.

nkaretnikov committed:
Update on "[primTorch] Minor improvements to doc and impl of `gaussian_nll_loss`"


Fixes #53392.

[ghstack-poisoned]

nkaretnikov commented on functorch/test/test_ops.py:
```diff
@@ -563,11 +563,8 @@ def fn(inp, *args, **kwargs):
         skip("nn.functional.feature_alpha_dropout", "with_train"),  # calls random op
         skip("nn.functional.fractional_max_pool2d"),  # calls random op
         skip("nn.functional.fractional_max_pool3d"),  # calls random op
+        skip("nn.functional.gaussian_nll_loss"),  # takes too long
```
As discussed offline, it's common for some ops to skip functorch tests. For the one that passes locally, I've added a note. The rest I just skip.

nkaretnikov committed:
Update on "[primTorch] Minor improvements to doc and impl of `gaussian_nll_loss`"


Fixes #53392.

[ghstack-poisoned]

nkaretnikov commented:
@lezcano @mruberry PTAL.

Summary:
- [removed no_grad](https://github.com/pytorch/pytorch/pull/85612#discussion_r984863143) from clamp and changed to functional clamp
- removed a call to `.item`, which resulted in better test coverage
- improved the docs to document allowed `var` inputs
- this doesn't add a ref (based on our discussion in another PR) because it's a pure Python impl.

mruberry commented on torch/nn/functional.py:
```diff
@@ -2762,6 +2762,7 @@ def poisson_nll_loss(
     return ret
 
 
+# TODO: Pure Python impl - don't add a primTorch ref
```
I would remove this comment -- it's not really a TODO since there's nothing to do

mruberry commented on torch/nn/functional.py:
```diff
@@ -2798,46 +2786,45 @@ def gaussian_nll_loss(
+    # If var is the same shape as input, it's the heteroscedastic case.
+    # If var is *not* the same shape as input, it's the homoscedastic case.
+    #
+    # To support broadcasting, the following sub-cases are allowed in the
```
It's not really broadcasting, though. We should be clear about what, exactly, it does, and not use a similar concept, which could confuse the reader

mruberry commented on torch/nn/modules/loss.py:
```diff
@@ -314,13 +314,13 @@ class GaussianNLLLoss(_Loss):
     size as ``input`` (due to a homoscedastic assumption), it must either have a final dimension
-    of 1 or have one fewer dimension (with all other sizes being the same) for correct broadcasting.
+    of 1 or have one fewer dimension (when comparing from the outermost dimension, with all other
+    sizes being the same) for correct later broadcasting.
```
Let's not use the term "broadcasting" here because this is not the same thing

mruberry commented on torch/nn/modules/loss.py:
```diff
@@ -314,13 +314,13 @@ class GaussianNLLLoss(_Loss):
     where :attr:`eps` is used for stability. By default, the constant term of
     the loss function is omitted unless :attr:`full` is ``True``. If ``var`` is not the same
     size as ``input`` (due to a homoscedastic assumption), it must either have a final dimension
```
I don't think the phrase "final dimension" is commonly understood

mruberry commented on torch/nn/modules/loss.py:
```diff
@@ -314,13 +314,13 @@ class GaussianNLLLoss(_Loss):
     the loss function is omitted unless :attr:`full` is ``True``. If ``var`` is not the same
     size as ``input`` (due to a homoscedastic assumption), it must either have a final dimension
-    of 1 or have one fewer dimension (with all other sizes being the same) for correct broadcasting.
+    of 1 or have one fewer dimension (when comparing from the outermost dimension, with all other
```
I would just break this into cases. I'm not sure the order of comparison makes it that clear. Either `var` has the same shape, has the same shape except its innermost dimension is 1, or has the same shape except it's "missing" the innermost dimension. In the last two cases the variance is assumed to be the same for each distribution (the distributions are assumed to be homoscedastic)

mruberry reviewed (commented):
The functional and test changes look good, but I think we should take the time to be more diligent with the docs. It could save us a lot of headaches in the future.

One option would be to separate the doc changes into another PR.

mruberry commented on torch/nn/functional.py:
```diff
@@ -2773,19 +2774,6 @@ def gaussian_nll_loss(
     r"""Gaussian negative log likelihood loss.
 
     See :class:`~torch.nn.GaussianNLLLoss` for details.
-
```
I would keep this.

While we do like to reduce document redundancy, in a more perfect world modules would just wrap functions, and the functions would be documented.
